services/route_manager/src/main/java/com/futurewei/alcor/route/controller/NeutronRouterController.java

```diff
@@ -324,12 +329,30 @@ public RoutesToNeutronWebResponse addRoutesToNeutronRouter(@PathVariable String
         RestPreconditionsUtil.verifyParameterNotNullorEmpty(routerid);
         RestPreconditionsUtil.verifyResourceFound(projectid);
 
-        RoutesToNeutronRouterRequestObject router = resource.getRouter();
-        if (router == null) {
+        NewRoutesWebRequest newRoutes = resource.getRouter();
+        if (newRoutes == null) {
             return new RoutesToNeutronWebResponse();
         }
 
-        RoutesToNeutronWebResponse routesToNeutronWebResponse = this.neutronRouterService.addRoutesToNeutronRouter(routerid, router);
+        // List<String> ports -> port entity -> subnet id
+        Router router = this.routerDatabaseService.getByRouterId(routerid);
+
+        RoutesToNeutronWebResponse routesToNeutronWebResponse = this.neutronRouterService.addRoutesToNeutronRouter(routerid, newRoutes);
+
+        List<String> gatewayPorts = router.getPorts();
+        List<String> subnetIds = this.routerToPMService.getSubnetIdsFromPM(projectid, gatewayPorts);
+        // sub-level routing rule update
+        List<InternalSubnetRoutingTable> internalSubnetRoutingTableList = new ArrayList<>();
+        for (String subnetId : subnetIds) {
+            UpdateRoutingRuleResponse updateRoutingRuleResponse = this.neutronRouterService.updateRoutingRule(subnetId, newRoutes, true);
+            List<InternalSubnetRoutingTable> internalSubnetRoutingTables = updateRoutingRuleResponse.getInternalSubnetRoutingTables();
+            internalSubnetRoutingTableList.addAll(internalSubnetRoutingTables);
+        }
+
+        InternalRouterInfo internalRouterInfo = this.neutronRouterService.constructInternalRouterInfo(internalSubnetRoutingTableList);
+
+        // send InternalRouterInfo contract to DPM
+        this.routerToDPMService.sendInternalRouterInfoToDPM(internalRouterInfo);
 
         // TODO:  l3-neighbors-updating (waiting for PM)
```

Comment:
We don't need todo for L3-neighbors here, you can remove it.

Reply:
NP